internal/services/comprehend.go

```diff
@@ -1,53 +1,106 @@
+    Result *comprehend.DetectEntitiesOutput
+    Error error
+}
+
+const ComprehendMaxChars = 5000
+
+func IsLetter(char uint8) bool {
+    return unicode.IsLetter(rune(char))
 }
 
 func GetEntitiesFromBodyText(bodyText string) ([]*comprehend.Entity, error) {
-	client, err := GetComprehendClient("developerPlayground")
-
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't create client")
-	}
-
-	input := &comprehend.DetectEntitiesInput{}
-	input.SetText(bodyText)
-	input.SetLanguageCode("en")
-	result, err := client.DetectEntities(input)
-	if err != nil {
-		return nil, errors.Wrap(err, "unable to get entities")
-	}
-	return result.Entities, nil
+    client, err := GetComprehendClient("developerPlayground")
+
+    if err != nil {
+        return nil, errors.Wrap(err, "couldn't create client")
+    }
+
+    // Use a separate goroutine to request each chunk, and wait for each to write to the channel
+    comprehendResults := make(chan ComprehendResult)
+    var wg sync.WaitGroup
+    wg.Add(int(math.Ceil( float64(len(bodyText)) / float64(ComprehendMaxChars) )))
+
+    for i := 0; i < len(bodyText); {
+        var end = i + ComprehendMaxChars-1
+
+        if end >= len(bodyText) {
+            //final chunk
+            end = len(bodyText)-1
+        } else if IsLetter(bodyText[end]) {
+            //Avoid splitting on a word
+            for j := end - 1; j >= i; j-- {
+                if !IsLetter(bodyText[j]) {
+                    end = j
+                    break
+                }
+            }
+        }
+
+        var chunk = bodyText[i:end]
+
+        go func(text string) {
+            defer wg.Done()
+
+            input := &comprehend.DetectEntitiesInput{}
+            input.SetText(text)
+            input.SetLanguageCode("en")
+            result, err := client.DetectEntities(input)
+            if err != nil {
+                fmt.Println("Comprehend request error", err)
+            }
+
+            comprehendResults <- ComprehendResult{result, err}
+        }(chunk)
+
+        i = end+1
+    }
+
+    go func() {
+        wg.Wait()
+        close(comprehendResults)
+    }()
+
+    results := make([]*comprehend.Entity, 0)
+    for response := range comprehendResults {
+        results = append(results, response.Result.Entities...)
+    }
+
+    return results, nil
 }
 
 func GetEntitiesFromPath(path string) ([]*comprehend.Entity, error) {
-	articleFields, err := GetArticleFieldsFromCapi(path, "test")
-	if err != nil {
-		return nil, errors.Wrap(err, "Couldn't get article fields from CAPI for given path")
-	}
-	var bodyText = articleFields.Fields.BodyText
-
-	// hack to stop it failing on long articles
-	if len(bodyText) > 4999 {
-		bodyText = articleFields.Fields.BodyText[0:4999]
-	}
-	entities, err := GetEntitiesFromBodyText(bodyText)
-
-	if err != nil {
-		return nil, errors.Wrap(err, "Error retrieving entities from body text")
-	}
-
-	return entities, nil
+    articleFields, err := GetArticleFieldsFromCapi(path, "test")
```

Comment:
we should pass in the api key here

Reply:
(i know this was already there, just noticed it!)

Reply:
oh that's weird, why are we even calling `GetArticleFieldsFromCapi` from here? It's called in `GetContentAnalysisForPath`!

Reply:
just pushed a commit to change this